### PR TITLE
chore: upgrade hbase to 2.5.3

### PR DESCRIPTION
**Pull request by igorbernstein2**
* Upgrade HBase 2 to 2.5.3 and fix all resulting build issues
* mutateRow() changed it's return type, so it needs to be pushed down into 1x/2x leaf modules. 
* mutateRow() now differs in HBase 2.3 and 2.4, so use runtime bytecode gen to handle the differences
* mutateRow() is now supposed to handle Append and Increment. Added support for the simple 1 mutation case. Bigtable can't handle multiple
* Table referenced shaded protos in its method signatures, so use bytecode gen to allow compilation
* Fix warnings about unclosed connections
* checkAndMutate was added to Table, left as a TODO
* mirroring-client now only supports hbase >= 2.4 due to mutateRow api changes. Left as a TODO, will require bytecode gen to work with earlier versions hbase. (Shitanshu approved)
* fixed IDE symbol lookup in mirroring client

### CODE REVIEW

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/AbstractBigtableTable.java [hunk 1]
```diff
@@ -501,11 +496,20 @@ private void mutateRow(Mutation mutation, RowMutation rowMutation, String type)
...
 
   /** {@inheritDoc} */
-  @Override
-  public void mutateRow(RowMutations rowMutations) throws IOException {
+  protected Result mutateRowBase(RowMutations rowMutations) throws IOException {
     LOG.trace("mutateRow(RowMutation)");
+
+    if (rowMutations.getMutations().size() == 1) {
+      Mutation onlyMutation = rowMutations.getMutations().get(0);
+      if (onlyMutation instanceof Append) {
```
Comment (mutianf) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/AbstractBigtableTable.java:504]: Can we add a comment saying that Bigtable only supports single append and increment, and line 516 is gonna reject the case where there are multiple append / increment ?

File: bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/BigtableTableCore.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+  @Override
+  public <T extends Service, R> Map<byte[], R> coprocessorService(
+      Class<T> aClass, byte[] bytes, byte[] bytes1, Batch.Call<T, R> call)
+      throws ServiceException, Throwable {
+    throw new UnsupportedOperationException("not implemented");
+  }
+
+  @Override
+  public <T extends Service, R> void coprocessorService(
```
Comment (mutianf) [inline bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/BigtableTableCore.java:49]: Why are only these coprocessorService method moved here?

File: bigtable-client-core-parent/bigtable-hbase/src/test/java/org/apache/hadoop/hbase/client/TestAbstractBigtableConnection.java [hunk 3]
```diff
@@ -262,7 +268,53 @@ protected SampledRowKeysAdapter createSampledRowKeysAdapter(
 
     @Override
     public Table getTable(TableName tableName, ExecutorService executorService) {
-      return new AbstractBigtableTable(this, createAdapter(tableName)) {};
+      return new AbstractBigtableTable(this, createAdapter(tableName)) {
```
Comment (mutianf) [inline bigtable-client-core-parent/bigtable-hbase/src/test/java/org/apache/hadoop/hbase/client/TestAbstractBigtableConnection.java:271]: why not call BigtableTableCore(..) here?

File: bigtable-dataflow-parent/bigtable-beam-import/pom.xml [hunk 4]
```diff
@@ -64,16 +64,24 @@ limitations under the License.
         <groupId>org.checkerframework</groupId>
         <artifactId>checker-qual</artifactId>
         <version>3.31.0</version>
+        <scope>provided</scope>
```
Comment (mutianf) [inline bigtable-dataflow-parent/bigtable-beam-import/pom.xml:67]: Please add a comment why we're marking them as provided

File: bigtable-hbase-1.x-parent/bigtable-hbase-1.x/src/main/java/com/google/cloud/bigtable/hbase1_x/BigtableTable.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+import com.google.cloud.bigtable.hbase.AbstractBigtableTable;
+import com.google.cloud.bigtable.hbase.adapters.HBaseRequestAdapter;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import net.bytebuddy.ByteBuddy;
+import org.apache.hadoop.hbase.client.AbstractBigtableConnection;
+import org.apache.hadoop.hbase.client.RowMutations;
+
+public abstract class BigtableTable extends AbstractBigtableTable {
```
Comment (mutianf) [inline bigtable-hbase-1.x-parent/bigtable-hbase-1.x/src/main/java/com/google/cloud/bigtable/hbase1_x/BigtableTable.java:26]: Do we need to annotate this with `@internalAPi`?

Comment (mutianf) [reply bigtable-hbase-1.x-parent/bigtable-hbase-1.x/src/main/java/com/google/cloud/bigtable/hbase1_x/BigtableTable.java:26]: And also maybe add some comments about what this class does?

File: bigtable-hbase-1.x-parent/bigtable-hbase-1.x/src/test/java/com/google/cloud/bigtable/hbase1_x/TestBigtableConnection.java [hunk 6]
```diff
@@ -94,11 +94,12 @@ public void tearDown() throws IOException {
...
   public void testOverloadedConstructor() throws IOException {
-    Connection hbaseConnection =
-        new BigtableConnection(configuration, false, Executors.newSingleThreadExecutor(), null);
+    try (Connection hbaseConnection =
+        new BigtableConnection(configuration, false, Executors.newSingleThreadExecutor(), null)) {
 
-    assertTrue(hbaseConnection.getAdmin() instanceof BigtableAdmin);
-    assertTrue(hbaseConnection.getTable(TABLE_NAME) instanceof AbstractBigtableTable);
+      assertTrue(hbaseConnection.getAdmin() instanceof BigtableAdmin);
+      assertTrue(hbaseConnection.getTable(TABLE_NAME) instanceof AbstractBigtableTable);
```
Comment (mutianf) [inline bigtable-hbase-1.x-parent/bigtable-hbase-1.x/src/test/java/com/google/cloud/bigtable/hbase1_x/TestBigtableConnection.java:101]: should this be `instanceof BigtableTable`?

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTable.java [hunk 7]
```diff
@@ -318,8 +338,18 @@ public CompletableFuture<Result> increment(Increment increment) {
 
   /** {@inheritDoc} */
   @Override
-  public CompletableFuture<Void> mutateRow(RowMutations rowMutations) {
-    return toCompletableFuture(clientWrapper.mutateRowAsync(hbaseAdapter.adapt(rowMutations)));
+  public CompletableFuture<Result> mutateRow(RowMutations rowMutations) {
+    if (rowMutations.getMutations().size() == 1) {
```
Comment (mutianf) [inline bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTable.java:342]: also add a comment here

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableTable.java [hunk 8]
```diff
@@ -249,4 +270,42 @@ public <R> void batchCallback(
...
+    return mutateRowBase(rowMutations);
+  }
+
+  private static Class<? extends BigtableTable> dynamicClass = null;
+
+  private static Class<? extends BigtableTable> getSubclass() throws NoSuchMethodException {
+    if (dynamicClass != null) {
+      return dynamicClass;
+    }
+    // default Result mutateRow(RowMutations rm) throws IOException {
```
Comment (mutianf) [inline bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableTable.java:288]: I think we can remove this comment?

File: hbase-migration-tools/mirroring-client/bigtable-hbase-mirroring-client-2.x-parent/bigtable-hbase-mirroring-client-2.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase2_x/MirroringAsyncTable.java [hunk 9]
```diff
@@ -522,6 +538,20 @@ public ResultScanner getScanner(Scan scan) {
         this.resultScannerBufferedMismatchedResults);
   }
 
+  @Override
+  public CompletableFuture<CheckAndMutateResult> checkAndMutate(CheckAndMutate checkAndMutate) {
+    CompletableFuture<CheckAndMutateResult> f = new CompletableFuture<>();
```
Comment (mutianf) [inline hbase-migration-tools/mirroring-client/bigtable-hbase-mirroring-client-2.x-parent/bigtable-hbase-mirroring-client-2.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase2_x/MirroringAsyncTable.java:543]: Add a todo here

Comment (igorbernstein2) [reply hbase-migration-tools/mirroring-client/bigtable-hbase-mirroring-client-2.x-parent/bigtable-hbase-mirroring-client-2.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase2_x/MirroringAsyncTable.java:544]: done

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/AbstractBigtableTable.java (hunk 1 above)
Comment (igorbernstein2) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/AbstractBigtableTable.java:508]: done

File: bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/BigtableTableCore.java (hunk 2 above)
Comment (igorbernstein2) [reply bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/BigtableTableCore.java:55]: added a comment

File: bigtable-client-core-parent/bigtable-hbase/src/test/java/org/apache/hadoop/hbase/client/TestAbstractBigtableConnection.java (hunk 3 above)
Comment (igorbernstein2) [reply bigtable-client-core-parent/bigtable-hbase/src/test/java/org/apache/hadoop/hbase/client/TestAbstractBigtableConnection.java]: oops, fixed

File: bigtable-dataflow-parent/bigtable-beam-import/pom.xml (hunk 4 above)
Comment (igorbernstein2) [reply bigtable-dataflow-parent/bigtable-beam-import/pom.xml:68]: done

File: bigtable-hbase-1.x-parent/bigtable-hbase-1.x/src/main/java/com/google/cloud/bigtable/hbase1_x/BigtableTable.java (hunk 5 above)
Comment (igorbernstein2) [reply bigtable-hbase-1.x-parent/bigtable-hbase-1.x/src/main/java/com/google/cloud/bigtable/hbase1_x/BigtableTable.java:28]: done

File: bigtable-hbase-1.x-parent/bigtable-hbase-1.x/src/test/java/com/google/cloud/bigtable/hbase1_x/TestBigtableConnection.java (hunk 6 above)
Comment (igorbernstein2) [reply bigtable-hbase-1.x-parent/bigtable-hbase-1.x/src/test/java/com/google/cloud/bigtable/hbase1_x/TestBigtableConnection.java]: done

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTable.java (hunk 7 above)
Comment (igorbernstein2) [reply bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTable.java:346]: done

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableTable.java (hunk 8 above)
Comment (igorbernstein2) [reply bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableTable.java]: oops

**Review by mutianf — APPROVED** [review]

**Pull request merged**